server/topdf/topdf.go

```diff
@@ -0,0 +1,143 @@
+package topdf
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"io/ioutil"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/plugin"
+)
+
+// toPDFPrefix used as a prefix while using KV store to access PDF files' fileID.
+const toPDFPrefix = "pdf:"
+
+// ErrUnauthorizedUser returned when user has no access to a file that requested to be converted to PDF.
+var ErrUnauthorizedUser = errors.New("user is not authorized to access pdf")
+
+// TOPDF is an application that converts files to PDFs and permanently caches them by using Mattermost APIs.
+type TOPDF struct {
+	// mapi is Mattermost's Plugin API.
+	mapi plugin.API
```

Comment:
Accepting the full `plugin.API` here seems to somewhat defeat the point of factoring out `topdf` as a server. Should we instead accept an interface with just the required functionality?

Reply:
I partially agree that it defeats the point of moving this logic to a separate package because as you said it still depends on the `plugin.API ` instead of a much more generic interface.

But still in my view, this abstraction builds some solid lines between the `main` pkg and actual `topdf` logic.

On the other hand, it's kind of hard to create a much more generic interface other than directly using `plugin.API` for `topdf` to accept because topdf heavily depends on `plugin.API`'s features.

That being said, I can still create an interface to only select the funcs that we need use from `plugin.API` as you pointed out but still this interface will be strongly tied to `plugin.API` since it'll be direct copy of its some funcs. But in a general perspective, having this interface seems better to be able to see what Plugin APIs that `topdf` package depends at the first look.

To get more precise about the design pattern I follow here,

I consider `plugin.API` as a `DAO` layer/provider here. `topdf` pkg is the actually app that uses DAO and I place the `main` pkg as a user interface layer to expose topdf's features to the network layer by using the hooks including ServeHTTP (-REST API handlers).

As a summary in the architectural point of view:
* `topdf` pkg is the actual app(controller) and uses `plugin.API` DAO to interact with an external API, in this case Mattermost server's features.

* `main` pkg is the network interface to expose features of `topdf` to other 3rd parties and also to configure `topdf` pkg to match with the needs of those 3rd parties.